### ISSUE TITLE
🏭 Implement Missing Tensor Factory Methods

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#15]
# 🏭 Implement Missing Tensor Factory Methods

## 📋 Description

Several tensor factory methods are declared but not implemented in the Tensor class. These methods are essential for creating common tensor patterns used in machine learning and scientific computing.

## 🎯 Acceptance Criteria

- [ ] Implement `Tensor.rand(*shape)` for uniform random tensors [0, 1)
- [ ] Implement `Tensor.eye(n, m=None)` for identity matrices
- [ ] Add `Tensor.empty(*shape)` for uninitialized tensors
- [ ] Add `Tensor.linspace(start, end, steps)` for linearly spaced values
- [ ] Add `Tensor.logspace(start, end, steps, base=10)` for logarithmically spaced values
- [ ] Implement `Tensor.normal(mean, std, shape)` as alias for randn with custom params
- [ ] Add comprehensive tests for all factory methods
- [ ] Ensure proper dtype and device support for all methods
- [ ] Add gradient support where applicable (requires_grad parameter)

### Current State
- `Tensor.ones()`: ✅ Implemented
- `Tensor.zeros()`: ✅ Implemented  
- `Tensor.full()`: ✅ Implemented
- `Tensor.arange()`: ✅ Implemented
- `Tensor.randn()`: ✅ Implemented
- `Tensor.rand()`: ❌ Stub only
- `Tensor.eye()`: ❌ Stub only
- `Tensor.empty()`: ❌ Not implemented
- `Tensor.linspace()`: ❌ Not implemented

### Implementation Requirements

**Random Uniform (rand)**
```python
@classmethod
def rand(cls, *shape: int, dtype: DTypeLike = dtypes.float32, 
         device: str = "cpu", requires_grad: bool = False) -> Tensor:
    """Create tensor with random values from uniform distribution [0, 1)"""
    size = _prod(shape) if shape else 1
    random_data = [random.random() for _ in range(size)]
    # Implementation similar to randn()
```

**Identity Matrix (eye)**
```python
@classmethod
def eye(cls, n: int, m: Optional[int] = None, dtype: DTypeLike = dtypes.float32,
        device: str = "cpu", requires_grad: bool = False) -> Tensor:
    """Create 2D tensor with ones on diagonal, zeros elsewhere"""
    m = m or n
    data = [[1.0 if i == j else 0.0 for j in range(m)] for i in range(n)]
    return cls(data, dtype=dtype, device=device, requires_grad=requires_grad)
```

**Linear Space (linspace)**
```python
@classmethod
def linspace(cls, start: float, end: float, steps: int, 
             dtype: DTypeLike = dtypes.float32, device: str = "cpu",
             requires_grad: bool = False) -> Tensor:
    """Create 1D tensor with linearly spaced values"""
    if steps < 2:
        raise ValueError("linspace requires at least 2 steps")
    step_size = (end - start) / (steps - 1)
    data = [start + i * step_size for i in range(steps)]
    return cls(data, dtype=dtype, device=device, requires_grad=requires_grad)
```

## 🧪 Testing Requirements

- [ ] Test all factory methods with different shapes
- [ ] Test with different dtypes (float32, float16, int32, etc.)
- [ ] Test edge cases (empty shapes, single elements)
- [ ] Test random methods produce expected distributions
- [ ] Test eye matrices have correct diagonal structure
- [ ] Test linspace produces correct sequences
- [ ] Test requires_grad propagation
- [ ] Performance benchmarks vs NumPy equivalents

## 📁 Files to Modify

- `grad/tensor.py` - Implement missing factory methods
- `tests/test_factory_methods.py` - Comprehensive test suite  
- `tests/test_random.py` - Random distribution tests
- Update docstrings and type hints

## 🔗 Dependencies

- Requires: Basic tensor creation (✅ implemented)
- Requires: Buffer system (✅ implemented)
- Uses: Existing `_filled()` helper method
- Blocks: Advanced tensor initialization patterns

### Usage Examples
```python
# Random uniform distribution
rand_tensor = Tensor.rand(3, 4, dtype=dtypes.float32)
print(rand_tensor)  # Values in [0, 1)

# Identity matrix
identity = Tensor.eye(5)
print(identity.shape)  # (5, 5)

# Non-square identity
rect_identity = Tensor.eye(3, 5)  # 3x5 with diagonal ones

# Linear spacing
linear = Tensor.linspace(0, 10, 11)  # [0, 1, 2, ..., 10]

# Logarithmic spacing  
log_vals = Tensor.logspace(0, 2, 3)  # [1, 10, 100]

# Empty tensor (uninitialized)
empty = Tensor.empty(2, 3, 4)  # Uninitialized values
```

### Integration with Existing Code
```python
# Should work with existing functionality
rand_matrix = Tensor.rand(100, 100, requires_grad=True)
identity = Tensor.eye(100, requires_grad=True)
result = rand_matrix @ identity  # Matrix multiplication
loss = result.sum()
loss.backward()  # Gradient computation
```

## 🚨 Edge Cases to Handle

- Zero-dimensional tensors for factory methods
- Invalid parameters (negative steps, invalid ranges)
- Memory constraints for large random tensors
- Numerical precision for linspace edge values
- Thread safety for random number generation
- Deterministic random generation (seed support)

## 🔧 Performance Considerations

- Efficient random number generation
- Memory allocation patterns
- Batch initialization for large tensors
- Consider future GPU random generation
- Minimize temporary memory usage

## 🏷️ Labels

- `enhancement`
- `factory-methods`
- `medium-priority`
- `good-first-issue`

## 📈 Priority

**Medium** - Important for usability but not blocking core functionality

## 🔗 Related Issues

- Enhances: Basic tensor creation
- Related to: Random number generation consistency
- Supports: Scientific computing use cases
